cmd/timescale-prometheus/main_test.go

```diff
@@ -0,0 +1,577 @@
+				response: c.readerResponse,
+				err:      c.readerErr,
+			}
+
+			handler := read(mockReader)
+
+			test := GenerateHandleTester(t, handler)
+
+			w := test("GET", getReader(c.requestBody))
+
+			if w.Code != c.responseCode {
+				t.Errorf("Unexpected HTTP status code received: got %d wanted %d", w.Code, c.responseCode)
+			}
+		})
+	}
+}
+
+func TestWrite(t *testing.T) {
+	testCases := []struct {
+		name             string
+		responseCode     int
+		requestBody      string
+		inserterResponse uint64
+		inserterErr      error
+		isLeader         bool
+		electionErr      error
+	}{
+		{
+			name:         "write request body error",
+			isLeader:     true,
+			responseCode: http.StatusInternalServerError,
+		},
+		{
+			name:         "malformed compression data",
+			isLeader:     true,
+			responseCode: http.StatusBadRequest,
+			requestBody:  "123",
+		},
+		{
+			name:         "malformed write request",
+			isLeader:     true,
+			responseCode: http.StatusBadRequest,
+			requestBody:  string(snappy.Encode(nil, []byte("test"))),
+		},
+		{
+			name:         "write error",
+			isLeader:     true,
+			responseCode: http.StatusInternalServerError,
+			inserterErr:  fmt.Errorf("some error"),
+			requestBody: writeRequestToString(
+				&prompb.WriteRequest{},
+			),
+		},
+		{
+			name:         "elector error",
+			electionErr:  fmt.Errorf("some error"),
+			responseCode: http.StatusOK,
+		},
+		{
+			name:         "not a leader",
+			responseCode: http.StatusOK,
+		},
+		{
+			name:             "happy path",
+			isLeader:         true,
+			responseCode:     http.StatusOK,
+			inserterResponse: 3,
+			requestBody: writeRequestToString(
+				&prompb.WriteRequest{
+					Timeseries: []prompb.TimeSeries{
+						{},
+					},
+				},
+			),
+		},
+	}
+
+	for _, c := range testCases {
+		t.Run(c.name, func(t *testing.T) {
+			elector = util.NewElector(
+				&mockElection{
+					isLeader: c.isLeader,
+					err:      c.electionErr,
+				},
+			)
+			mockGauge := &mockGauge{}
+			leaderGauge = mockGauge
+			mock := &mockInserter{
+				result: c.inserterResponse,
+				err:    c.inserterErr,
+			}
+
+			handler := write(mock)
+
+			test := GenerateHandleTester(t, handler)
+
+			w := test("GET", getReader(c.requestBody))
+
+			if w.Code != c.responseCode {
+				t.Errorf("Unexpected HTTP status code received: got %d wanted %d", w.Code, c.responseCode)
+			}
+
+			if c.electionErr != nil && mockGauge.value != 0 {
+				t.Errorf("leader gauge metric not set correctly: got %f when election returns an error", mockGauge.value)
+			}
+
+			switch {
+			case c.isLeader && mockGauge.value != 1:
+				t.Errorf("leader gauge metric not set correctly: got %f when is leader", mockGauge.value)
+			case !c.isLeader && mockGauge.value != 0:
+				t.Errorf("leader gauge metric not set correctly: got %f when is not leader", mockGauge.value)
+			}
+		})
+	}
+}
+
+func TestInitElector(t *testing.T) {
+	// TODO: refactor the function to be fully testable without using a DB.
```

Comment:
can we set this up as a integration test for now?

Reply:
The reason we can't use DB for testing is because main package can't access internal packages. Otherwise, I would have added it here. I'm unsure how to resolve it right now so I'm leaving it for the future. Ideally, we should refactor the code to be more testable vs. just using integration tests and spinning up docker containers for every test.